File: 11,12,14/boj9663.py
n = int(input())
cnt = 0
cols = [-1]*n

def isPossible(lv):
    for i in range(lv):
        #같은 열에 위치
        if cols[i] == cols[lv]:
            return False
        #대각선에 위치
        elif abs(cols[i]-cols[lv]) == abs(i-lv):
            return False
    return True

def nQueen(lv):
    global cnt
    if lv == n:
        cnt += 1
        return 0
    else: #현재 lv에서
        for i in range(n): #가능한 열 위치
            cols[lv] = i
            if isPossible(lv):
                nQueen(lv+1)
nQueen(0)
print(cnt)

File: 11,12,14/boj9663.c
#include <stdio.h>
#include <stdlib.h>
int n;
int cnt = 0;
int cols[15] = {0, };

int isPossible(int lv){
    for (int i = 0; i<lv; i++){
        if (cols[i] == cols[lv]) return 0;//같은 열
        else if (abs(cols[i]-cols[lv]) == abs(i-lv)) return 0;//대각선
    }
    return 1;
}
void nQueen(int lv){
    if (lv == n){
        cnt++;
        return;
    }
    else{ //현재 lv에서
        for (int i = 0; i<n; i++){ //가능한 열 위치
            cols[lv] = i;
            if (isPossible(lv)) nQueen(lv+1);
                         }
        }
}                   
int main(){
    scanf("%d", &n);
    nQueen(0);
    printf("%d", cnt);
    return 0;
}